drv/opentitan-aontimer/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+
+#[allow(unused_imports)]
+use userlib::*;
+// use zerocopy::AsBytes;
+
+use core::ptr::slice_from_raw_parts_mut;
+
+const WDOG_REGWEN: usize = 4;
+const WDOG_CTRL: usize = 5;
+const WDOG_BARK_THOLD: usize = 6;
+const WDOG_BITE_THOLD: usize = 7;
+const WDOG_COUNT: usize = 8;
+const INTR_STATE: usize = 9;
+
+// The OpenTitan Always-On Timer is a counter that increments. If the counter
+// reaches a configurable threshold, it will emit an interrupt signal, called the
+// "bark". If the counter reaches a second configurable threshold, it will reset
+// the system, called the "bite".
+// Enabling the AON Timer before setting a bite threshold causes the system to
+// immediately reset.
+pub struct AonTimer {
+    base_addr: *mut [u32],
+    clock_freq: u32,
+    pub bark_cb: Option<fn()>,
+}
+
+impl AonTimer {
+    // Initialize a new watchdog driver device with the specified base address,
+    // clock frequency (in Hz), and the bark/bite thresholds (both in ms).
+    // If the bark threshold > bite threshold, the bite threshold is set to
+    // the bark threshold.
+    pub fn new(
+        base_addr: u32,
+        clock_freq: u32,
+        bark_thold_ms: u64,
+        bite_thold_ms: u64,
+        bark_cb: Option<fn()>,
+    ) -> Self {
+        let inst = AonTimer {
+            base_addr: slice_from_raw_parts_mut(base_addr as *mut u32, 12),
+            clock_freq,
+            bark_cb,
+        };
+        unsafe {
+            //1: init regs
+            (*inst.base_addr)[WDOG_COUNT] = 0;
+            (*inst.base_addr)[INTR_STATE] = 0b11;
```

Comment:
Can you please define these register fields instead of using magic bits?

Reply:
WIll change.  This does raise the question if we should use libraries such as[ tock-registers](https://crates.io/crates/tock-registers).  Probably not needed here, but we should decide if we recommend it or not as people start developing more tasks.  I know @FawazTirmizi has done some research on this.  I think there were code size concerns?

Reply:
After reading the code, I had the same question in my mind. If we should bring in the tock-register libraries for cleaner register field access. @FawazTirmizi - do you have the code size concerns captured somewhere? If not, can you please start a doc or bug to start tracking this activity?

Reply:
I just created some `const`.  Don't think it warranted adding tock.